SIP-X: The Sovryn Improvement Proposal Process

john-light opened the pull request:
Edit: This PR has been closed and superseded by https://github.com/DistributedCollective/SIPS/pull/11

john-light committed:
SGP-0001 v1

john-light committed:
Fix typos

john-light committed:
Fix typo

john-light committed:
Add info about voting power

YagoBit commented:
This is a great start! Having a review process is crucial. Some thoughts:
1. IP has the advantage of being easy to pronounce
2. I don't like the idea of having a multi-sig gate the ability to offer proposals. This make the system far less permissionless.
3. That said, having a culture of rejecting proposals that do not go through the review process by default does make sense.
4. Love the ideas of having categories. Eg. SIP-P0004 could be proclamation. SIP-T0003 could be treasury. SIP-C0004 might be change to contract. I haven't given enough thought to what the categories might b. However, some SIP introduce no changes on state, and are more cultural/conceptual, and that should probably be reflected. Your proposed tracks seem well thought out. 
5. I think having the ability to have community poll, or off-chain poll of stakers can help smooth the process of prioritising proposals and avoiding controversial ones. I have seen snapshot governance tools used for this purpose. 
6. To me it seems that the most obvious candidates for Guardians are delegates. 
7. In any event, we should seek to make sure that debate is always help transparently and publicly.

john-light commented:
> IP has the advantage of being easy to pronounce

Fair :slightly_smiling_face: Will change back to keep the current terminology.

> I don't like the idea of having a multi-sig gate the ability to offer proposals. This make the system far less permissionless.

Yes this was taking the Guardians concept and leaning hard into it. We can ease up and leave them only with veto power if the community prefers.

> That said, having a culture of rejecting proposals that do not go through the review process by default does make sense.

Sure, we can rely on social norms instead of a hard gating function.

> However, some SIP introduce no changes on state, and are more cultural/conceptual, and that should probably be reflected.

I think these kinds of proposals most likely belong in the Meta track or Proclamations track (the Sovryn Constitution isn't written yet but would be modified using a Meta track proposal). However the process does specify the possibility of using "Other" as a category and a way to add other categories as needed if a proposal doesn't fit into any existing category.

> I think having the ability to have community poll, or off-chain poll of stakers can help smooth the process of prioritising proposals and avoiding controversial ones. I have seen snapshot governance tools used for this purpose.

True yes. We could implement this as part of the social norms, similar to the review process: proposals are reviewed, then polled, then go back for a review if they fail the poll or proceed to the final vote if they pass the poll. Or we can say it's optional that proposals go through a poll before the final vote. Either way I do agree having a polling system would be useful.

> To me it seems that the most obvious candidates for Guardians are delegates.

Right now I have it that they are elected by SOV holders, and there are a limited number of them (5). Whereas there can be an unlimited number of delegates, and some holders (large and small) won't delegate at all.

However I've been thinking also that it might be better to have the Guardians elected by some other means. The thinking here is that the purpose of Guardians is to protect the project against malicious proposals. Proposals that would be approved by... SOV holders. And who elects the Guardians? SOV holders...? So if there is an attacker, they could just elect Guardians who won't reject their proposals. To make this system robust Guardians should instead be elected by some other means. Perhaps initially selected by the team, and then self-perpetuating into the future (i.e. the Guardians themselves vote on any replacements that are needed if any of them leave). As a backstop against malicious/corrupt Guardians (e.g. if they go rogue and reject all proposals unless they receive some exorbitant rent, or are pressured by some external entity to block proposals) and assuming a majority of SOV holders are honest, the protocol could revert to pure SOV governance if no proposals are approved for X amount of time. This gives time for anyone who isn't comfortable participating anymore to sell out of the protocol, and time for the remaining honest SOV holders to coordinate a replacement for the Guardians.

> In any event, we should seek to make sure that debate is always help transparently and publicly.

Absolutely :slightly_smiling_face:

john-light committed:
Update SGP -> SIP

john-light committed:
Make changes based on feedback

- Biggest change was making most of the process optional  
- Changed the process so that anyone can submit proposals for a vote and Guardians only veto after a proposal has been submitted for a vote

john-light commented:
@YagoBit I made a few changes based on your feedback. Biggest change was making most of the process optional (so proposals can be accelerated if needed). And I changed the process so that anyone can submit proposals for a vote and Guardians only veto after a proposal has been submitted for a vote. Still thinking about changing how Guardians are selected, as mentioned in my comment above.

YagoBit commented:
While the idea of guardians makes me somewhat uncomfortable, I see it's utility - at least in the short term. Perhaps we should allow a supermajority to overrule the guardians?

jjmr007 commented:
We can learn some valuable things from how EIPs are processed... They use thair [html-proofer](https://rubygems.org/gems/html-proofer) and their [eip-validator](https://rubygems.org/gems/eip_validator)...

ororopickpocket commented on SIPs/SIP-000X.md:
```diff
@@ -0,0 +1,164 @@
+In addition to the requirement that all SIPs must be consistent with SIP-0, each track has its own requirements for SIPs as follows:
+
+**Contract**  
+Proposals made to the Contract track MUST change one or more of the Sovryn smart contracts. The proposal MUST link to the existing contract(s) to be modified and to the source code of the proposed contract modification(s). All Contract track proposals MUST include a link to a third party audit report of the specified modification(s).
```
I agree that all of our code must be audited by a third party. We've got a third party involved on an ongoing basis. All of our PRs are getting reviewed before they are deployed. At this moment, we do not have audit reports for these ongoing reviews, so we can't add a link to one. Should it suffice to send a link to the PRs, so people can verify that the changes have been sufficiently reviewed and approved?

YagoBit replied:
Why don't we have audit reports?

ororopickpocket replied:
i guess it's to keep the overhead low. if it's needed we can start requesting a report for each review. i'm just stating the current circumstances.

john-light replied:
> Should it suffice to send a link to the PRs, so people can verify that the changes have been sufficiently reviewed and approved?

@ororopickpocket Ultimately it's up to each voter to decide how much or how little information they want to see in each proposal. Right now there's not really a way to enforce this process, only to strongly suggest that it be followed. So while the intention here is that each Contract proposal have a proper audit report, voters may adjust their expectations for certain proposals and we can't force anything otherwise.

If it helps contextualize this requirement, the reason I propose a formal audit report is that I expect contract changes to be few and far between once the system gets to a certain point of stability. So while repos may have a lot of activity, changes won't be merged onchain right away but could/would be applied in batches. With this, having a proper audit report for each Contract proposal shouldn't be too much overhead. Open to suggested changes here though, as this is a bit outside my area of expertise!

ororopickpocket commented on SIPs/SIP-000X.md:
```diff
@@ -0,0 +1,164 @@
+
+After a SIP has gone through its Final Review, a SIP Editor will add a "Ready to vote" tag to the pull request. At the soonest available opportunity, Sovryn Guardians will review proposals that have the "Ready to vote" tag and decide whether or not they will veto the proposal.
+
+Once a proposal has been tagged "Ready to vote", the SIP author MAY submit the proposal to the Sovryn Governance Dashboard for a vote by Sovryn Voters. Anyone else MAY submit the proposal instead but they SHOULD make an effort to communicate with the SIP author first to make sure the proposal is ready to be submitted and will not be accidentally submitted multiple times.
```
the SIP author may not be able to create the proposal himself, because proposal creation is limited to addresses with at least 1% (delegated) voting power. Therefore, someone with sufficient voting power should be in charge of doing it.

john-light replied:
@ororopickpocket Yes I tried to keep this part worded somewhat vaguely so as to not be overly prescriptive regarding how proposals are submitted. I deliberately chose not to mention the parameters of the proposal submission requirements (or the parameters of the vote themselves) since these are defined elsewhere and could change -- and I see no need to have to update this document every time there's a simple parameter change on the governance contracts. 

That said, you bring to mind a good idea, which is that we may want to point to the governance contract documentation somewhere in this document (perhaps in this section you're commenting on) once it's ready, to at least let proposal authors reading this know that they should be aware of the governance contract requirements for submitting and passing a proposal onchain.

ororopickpocket reviewed (commented):
Nice

john-light committed:
Add issuance to proposal tracks list

john-light committed:
Fix typo

john-light committed:
soften language

john-light committed:
Fix typo

ororopickpocket commented on SIPs/SIP-000X.md:
```diff
@@ -0,0 +1,161 @@
+**Templates**  
+Below is a list of SIP templates for each track. Copy the template for the track your SIP is in, fill it out, and submit the pull request with your SIP for review. Sections marked as “required” in the template MUST be completed. Note that all SIPs MUST be licensed CC-0.
+
+* [Contract](../templates/contract_template.md)  
+* [Finance](../templates/finance_template.md)  
+* [Issuance](../templates/issuance_template.md)  
+* [Meta](../templates/meta_template.md)  
+* [Proclamations](../templates/proclamation_template.md)  
```
these links are not working here

john-light replied:
@ororopickpocket yes, because I moved the SIP PR here: https://github.com/DistributedCollective/SIPS/pull/11 and the template files here: https://github.com/DistributedCollective/SIPS/pull/10

This PR has been closed and superseded by https://github.com/DistributedCollective/SIPS/pull/11

tjcloa commented on SIPs/SIP-000X.md:
```diff
@@ -0,0 +1,161 @@
+
+There are five tracks that a SIP can be categorized into. Select the one you think is best for your SIP:
+
+* Contract: proposals for modifying one or more Sovryn smart contracts  
+* Finance: proposals for transferring funds from the Sovryn Treasury  
+* Issuance: proposals for increasing the total supply of SOV using the SOV token contract
+* Meta: proposals for changing SIP-X (“changing the way things are changed”)  
+* Proclamations: proposals for making a public statement on behalf of Sovryn  
```
can't proposals span over several tracks?

john-light replied:
@tjcloa can you give an example?
